packages/react-ui/src/hooks/useImperativeIntl.js

```diff
@@ -9,21 +9,36 @@ import {
 
 const cache = createIntlCache();
 
-export default function useImperativeIntl(intlConfig) {
-  return useMemo(() => {
+let lastIntlConfig;
+let cachedC4rIntl;
+
+const getGloballyCachedIntl = (intlConfig) => {
+  if (!cachedC4rIntl || lastIntlConfig !== intlConfig) {
+    // This is very simple cache exploits fact that Intl instance is actually same for most of time
+    // so we can reuse those maps across several instances of same components
+    // note, useMemo can't do that globally and flattenMessages over _app_ and c4r is quite costly and would
+    // be paid for every c4r component mounted.
     const locale = intlConfig?.locale || DEFAULT_LOCALE;
     const messagesLocale = findMatchingMessagesLocale(locale, messages);
     const intMessages = {
       ...(messages[messagesLocale] || {}),
       ...(intlConfig?.messages || {})
     };
 
-    return createIntl(
+    const combinedMessages = flattenMessages(intMessages);
+    cachedC4rIntl = createIntl(
       {
         locale,
-        messages: flattenMessages(intMessages)
+        messages: combinedMessages
       },
       cache
     );
-  }, [intlConfig]);
+    lastIntlConfig = intlConfig;
+  }
+  return cachedC4rIntl;
+};
+
+export default function useImperativeIntl(intlConfig) {
+  // second level cache, in components is just to avoid re-creating the Intl instance if user is rendering many languages in one app
+  return useMemo(() => getGloballyCachedIntl(intlConfig), [intlConfig]);
```

Comment:
The second-level cache only matters if we're rendering multiple languages _simultaneously_, right? If we just have a language dropdown, and switch from one language to another, the primary cache should be enough? 

I wonder if a single global cache of `WeakMap<IntlConfig, Intl>` would be simpler than two caches? I'm happy with either, though. I assume using multiple languages simultaneously is rare (maybe just SSR?).

Reply:
I've backed off to WeakMap based solution.
This weakmap will mostly have one entry so shouldn't be slower than useMemo.

@donmccurdy Can you take another view?